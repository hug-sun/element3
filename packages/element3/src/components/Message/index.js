export { Message } from './src/Message.js'
