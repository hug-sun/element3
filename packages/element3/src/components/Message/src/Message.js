import messageComponent from './Message.vue'
import { createComponent } from '../../../composables/component'
import { PopupManager } from '../../../utils/popup'
import { isVNode } from 'vue'

const instanceList = []

export function Message(opts) {
  return createMessage(mergeOptions(opts))
}

Message.closeAll = () => {
  instanceList.forEach((instance) => {
    instance.proxy.close()
    removeInstance(instance)
  })
}
;['info', 'success', 'warning', 'error'].forEach((type) => {
  Message[type] = (opts) => {
    return createMessage(mergeOptions(opts, type))
  }
})

function createMessage(opts) {
  const instance = createMessageComponentByOpts(opts)
  setZIndex(instance)
  appendToBody(instance)
  addInstance(instance)
  return instance.proxy
}

function createMessageComponentByOpts(opts) {
  if (isVNode(opts.message)) {
    return createComponent(messageComponent, opts, () => opts.message)
  }
  return createComponent(messageComponent, opts)
}

function setZIndex(instance) {
  instance.vnode.el.style.zIndex = PopupManager.nextZIndex()
}

function mergeOptions(opts, type = 'info') {
  const defaultOptions = {
    duration: 4500,
    type,
    offset: calculateVerticalOffset(opts.offset)
  }

  const userOnClose = opts?.onClose
  // opts.onClose Cannot be merged into the default options
  delete opts?.onClose
  delete opts?.offset
  defaultOptions.onClose = (instance) => {
    closeMessage(instance)
    if (userOnClose) userOnClose(instance.proxy)
  }

  if (typeof opts === 'string' || isVNode(opts)) {
    defaultOptions.message = opts
    return defaultOptions
  }

  return Object.assign({}, defaultOptions, opts)
}

function calculateVerticalOffset(offset = 20) {
  let result = offset

  instanceList.forEach((instance) => {
    result += getNextElementInterval(instance)
  })

  return result
}

function closeMessage(instance) {
  updatePosition(instance)
  removeInstance(instance)
}

function updatePosition(closeInstance) {
  const currentInstanceIndex = getIndexByInstance(closeInstance)

  if (currentInstanceIndex < 0) return

  for (
    let index = currentInstanceIndex + 1;
    index < instanceList.length;
    index++
  ) {
    const instance = instanceList[index]
    instance.proxy.offsetVal -= getNextElementInterval(closeInstance)
  }
}

function getNextElementInterval(instance) {
  const INTERVAL_HEIGHT = 16
  return instance.vnode.el.offsetHeight + INTERVAL_HEIGHT
}

function addInstance(instance) {
  instanceList.push(instance)
}

function removeInstance(instance) {
  instanceList.splice(getIndexByInstance(instance), 1)
}

function getIndexByInstance(instance) {
  return instanceList.findIndex((i) => i.uid == instance.uid)
}

function appendToBody(componentInstance) {
  document.body.append(componentInstance.vnode.el)
}
