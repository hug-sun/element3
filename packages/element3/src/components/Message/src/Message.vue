<template>
  <transition name="el-message-fade" @after-leave="handleAfterLeave" appear>
    <div
      v-show="isShow"
      class="el-message"
      :class="[
        isShowType ? `el-message--${type}` : '',
        showClose ? 'is-closable' : '',
        center ? 'is-center' : '',
        customClass
      ]"
      :style="[positionStyle]"
      @mouseenter="handleMouseenter"
      @mouseleave="handleMouseleave"
    >
      <i v-if="iconClass" :class="iconClass"></i>
      <i v-else :class="['el-message__icon', `el-icon-${type}`]"></i>
      <slot>
        <p
          class="el-message__content"
          v-if="dangerouslyUseHTMLString"
          v-html="message"
        ></p>
        <p class="el-message__content" v-else>
          {{ message }}
        </p>
      </slot>
      <i
        v-if="showClose"
        class="el-message__closeBtn el-icon-close"
        @click="handleClose"
      ></i>
    </div>
  </transition>
</template>

<script>
import { getCurrentInstance, computed, ref } from 'vue'
export default {
  props: {
    message: {
      type: [String, Object]
    },
    type: {
      type: String,
      defalut: 'info',
      validator(val) {
        return ['success', 'warning', 'info', 'error'].includes(val)
      }
    },
    iconClass: String,
    showClose: Boolean,
    duration: Number,
    center: Boolean,
    customClass: String,
    dangerouslyUseHTMLString: Boolean,
    verticalOffset: Number
  },
  emits: ['close'],
  setup(props, { emit }) {
    const instance = getCurrentInstance()

    const isShow = ref(true)
    // @public
    const verticalOffsetVal = ref(props.verticalOffset)

    const isShowType = computed(() => props.type && !props.iconClass)
    const positionStyle = computed(() => ({
      top: `${verticalOffsetVal.value}px`
    }))

    let timer
    function delayClose() {
      if (props.duration > 0) {
        timer = setTimeout(() => {
          _close()
        }, props.duration)
      }
    }

    function _close() {
      clearTimeout(timer)
      emit('close', instance)
      isShow.value = false
    }

    function handleAfterLeave() {
      instance.vnode.el.parentElement?.removeChild(instance.vnode.el)
    }

    function handleClose() {
      _close()
    }

    function handleMouseenter() {
      clearTimeout(timer)
    }

    function handleMouseleave() {
      delayClose()
    }

    // @public
    function close() {
      _close()
    }

    delayClose()

    return {
      close,
      isShow,
      isShowType,
      positionStyle,
      verticalOffsetVal,
      handleClose,
      handleAfterLeave,
      handleMouseenter,
      handleMouseleave
    }
  }
}
</script>

<style></style>
