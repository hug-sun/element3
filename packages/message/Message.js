import { defineComponent, createVNode, render, isVNode } from 'vue'
import Main from './Message.vue'
import { PopupManager } from '../../src/utils/popup'
import merge from '../../src/utils/merge'

const MessageConstructor = defineComponent(Main)

let instance
const instances = []
let seed = 1

const Message = function (options) {
  // if (Vue.prototype.$isServer) return
  // check ssr
  if (window === undefined) return

  options = options || {}
  if (typeof options === 'string' || isVNode(options)) {
    options = {
      message: options
    }
  }
  const userOnClose = options.onClose
  const id = 'message_' + seed++

  options.onClose = function () {
    Message.close(id, userOnClose)
  }
  let verticalOffset = options.offset || 20
  instances.forEach((item) => {
    verticalOffset += item.el.offsetHeight + 16
  })

  const nextZIndex = PopupManager.nextZIndex()
  options = merge(
    {
      id,
      verticalOffset,
      zIndex: nextZIndex
    },
    options
  )
  instance = createVNode(
    MessageConstructor,
    options,
    isVNode(options.message)
      ? {
          default: () => options.message
        }
      : null
  )

  const container = document.createElement('div')
  render(instance, container)
  instance.el.style.zIndex = nextZIndex
  document.body.appendChild(instance.el)
  instances.push(instance)
  return instance
}

;['success', 'warning', 'info', 'error'].forEach((type) => {
  Message[type] = (options) => {
    if (typeof options === 'string' || isVNode(options)) {
      options = {
        message: options
      }
    }
    options.type = type
    return Message(options)
  }
})

Message.close = function (id, userOnClose) {
  const len = instances.length
  let index = -1
  let removedHeight
  for (let i = 0; i < len; i++) {
    if (id === instances[i].component.props.id) {
      removedHeight = instances[i].el.offsetHeight
      index = i
      if (typeof userOnClose === 'function') {
        userOnClose(instances[i])
      }
      instances.splice(i, 1)
      break
    }
  }
  if (len <= 1 || index === -1 || index > instances.length - 1) return
  for (let i = index; i < len - 1; i++) {
    const dom = instances[i].el
    dom.style.top = parseInt(dom.style.top, 10) - removedHeight - 16 + 'px'
  }
}

Message.closeAll = function () {
  for (let i = instances.length - 1; i >= 0; i--) {
    instances[i].component.ctx.close()
  }
}

export default Message
