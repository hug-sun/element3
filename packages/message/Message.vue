<template>
  <transition name="el-message-fade" @after-leave="handleAfterLeave" appear>
    <div
      :class="[
        'el-message',
        type && !iconClass ? `el-message--${type}` : '',
        center ? 'is-center' : '',
        showClose ? 'is-closable' : '',
        customClass
      ]"
      :style="positionStyle"
      v-show="visible"
      @mouseenter="clearTimer"
      @mouseleave="startTimer"
      role="alert"
    >
      <i :class="iconClass" v-if="iconClass"></i>
      <i :class="typeClass" v-else></i>
      <slot>
        <p v-if="!dangerouslyUseHTMLString" class="el-message__content">
          {{ message }}
        </p>
        <p v-else v-html="message" class="el-message__content"></p>
      </slot>
      <i
        v-if="showClose"
        class="el-message__closeBtn el-icon-close"
        @click="close"
      ></i>
    </div>
  </transition>
</template>

<script type="text/babel">
import { computed, reactive, watch, onMounted, onUnmounted, toRefs } from 'vue'

const typeMap = {
  success: 'success',
  info: 'info',
  warning: 'warning',
  error: 'error'
}

export default {
  props: {
    id: {
      type: String,
      default: ''
    },
    message: {
      type: [String, Object],
      default: ''
    },
    duration: {
      type: Number,
      default: 3000
    },
    type: {
      type: String,
      default: 'info',
      validator(val) {
        return ['success', 'warning', 'info', 'error'].indexOf(val) > -1
      }
    },
    iconClass: {
      type: String,
      default: ''
    },
    customClass: {
      type: String,
      default: ''
    },
    showClose: {
      type: Boolean,
      default: false
    },
    verticalOffset: {
      type: Number,
      default: 20
    },
    dangerouslyUseHTMLString: {
      type: Boolean,
      default: false
    },
    center: {
      type: Boolean,
      default: false
    },
    onClose: {
      type: Function,
      required: true
    }
  },
  setup(props) {
    const state = reactive({
      visible: true,
      closed: false,
      timer: null
    })

    const typeClass = computed(() => {
      return props.type && !props.iconClass
        ? `el-message__icon el-icon-${typeMap[props.type]}`
        : ''
    })

    const positionStyle = computed(() => ({ top: `${props.verticalOffset}px` }))

    const close = () => {
      state.closed = true
      if (typeof props.onClose === 'function') {
        props.onClose(this)
      }
    }

    const clearTimer = () => {
      clearTimeout(state.timer)
    }

    const startTimer = () => {
      if (props.duration > 0) {
        state.timer = setTimeout(() => {
          if (!state.closed) {
            close()
          }
        }, props.duration)
      }
    }

    const keydown = (e) => {
      if (e.keyCode === 27) {
        // esc关闭消息
        if (!state.closed) {
          props.onClose(this)
        }
      }
    }

    const handleAfterLeave = (currentElement) => {
      if (currentElement && currentElement.parentNode) {
        currentElement.parentNode.removeChild(currentElement)
      }
    }

    watch(
      () => state.closed,
      (newVal) => {
        if (newVal) {
          state.visible = false
        }
      }
    )

    onMounted(() => {
      startTimer()
      document.addEventListener('keydown', keydown)
    })

    onUnmounted(() => {
      document.removeEventListener('keydown', keydown)
    })

    return {
      typeClass,
      positionStyle,
      close,
      clearTimer,
      startTimer,
      handleAfterLeave,
      ...toRefs(state)
    }
  }
}
</script>
