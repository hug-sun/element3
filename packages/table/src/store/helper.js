import Store from './index'
import { debounce } from 'lodash'

export function createStore(table, initialState = {}) {
  if (!table) {
    throw new Error('Table is required.')
  }

  const store = Store()
  // fix https://github.com/ElemeFE/element/issues/14075
  // related pr https://github.com/ElemeFE/element/pull/14146
  store.toggleAllSelection = debounce(store._toggleAllSelection, 10)
  Object.keys(initialState).forEach((key) => {
    store.states[key].value = initialState[key]
  })
  return store
}

export function mapStates(mapper) {
  const res = {}
  Object.keys(mapper).forEach((key) => {
    const value = mapper[key]
    let fn
    if (typeof value === 'string') {
      fn = function () {
        return this.store.states[value]
      }
    } else if (typeof value === 'function') {
      fn = function () {
        return value.call(this, this.store.states)
      }
    } else {
      console.error('invalid value type')
    }
    if (fn) {
      res[key] = fn
    }
  })
  return res
}
