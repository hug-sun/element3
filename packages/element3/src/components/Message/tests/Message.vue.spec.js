import { mount, flushPromises } from '@vue/test-utils'
import Message from '../src/Message.vue'
import { h } from 'vue'
import { expectHaveStyle } from '../../../../tests/helper'

describe('Message.vue', () => {
  test('snapshot', () => {
    const wrapper = mount(Message)
    expect(wrapper.element).toMatchSnapshot()
  })
  test('message', () => {
    const message = 'this is a message'

    const wrapper = mount(Message, {
      props: {
        message
      }
    })

    expect(wrapper.get('.el-message__content').text()).toContain(message)
  })

  test('message is vnode', () => {
    const message = h('p', { class: 'el-message__content' }, 'foo')

    const wrapper = mount(Message, {
      slots: {
        default: message
      }
    })

    expect(wrapper.get('.el-message__content').html()).toContain(
      '<p class="el-message__content">foo</p>'
    )
  })

  describe('type', () => {
    test('show type icon', () => {
      const type = 'info'
      const wrapper = mount(Message, {
        props: {
          type
        }
      })

      expect(wrapper.find(`.el-icon-${type}`).exists()).toBe(true)
      expect(wrapper.find('.el-message__icon').exists()).toBe(true)
      expect(wrapper.get('.el-message').classes()).toContain(`el-message--info`)
    })
  })

  test('iconClass', () => {
    const wrapper = mount(Message, {
      props: {
        iconClass: 'el-icon-setting',
        type: 'info'
      }
    })

    expect(wrapper.find('.el-icon-setting').exists()).toBe(true)
  })

  test('should just icon is exists ', () => {
    const wrapper = mount(Message, {
      props: {
        iconClass: 'el-icon-setting',
        type: 'info'
      }
    })

    expect(wrapper.find('.el-icon-setting').exists()).toBe(true)
    expect(wrapper.find('.el-icon-info').exists()).toBe(false)
    expect(wrapper.get('.el-message').classes()).not.toContain(
      'el-message--info'
    )
  })

  describe('close', () => {
    test('showClose ', () => {
      const wrapper = mount(Message, {
        props: {
          showClose: true
        }
      })

      expect(wrapper.find('.el-message__closeBtn').exists()).toBe(true)
      expect(wrapper.find('.el-icon-close').exists()).toBe(true)
      expect(wrapper.get('.el-message').classes()).toContain('is-closable')
    })

    test('click close btn', async () => {
      const wrapper = mount(Message, {
        props: {
          showClose: true
        }
      })

      const closeBtn = wrapper.get('.el-message__closeBtn')
      await closeBtn.trigger('click')
      expect(wrapper.get('.el-message').isVisible()).toBe(false)
    })

    test('duration: message is closed when in 1000ms', async () => {
      jest.useFakeTimers()

      const wrapper = mount(Message, {
        props: {
          duration: 1000
        }
      })
      jest.runTimersToTime(1000)
      await flushPromises()
      expect(wrapper.get('.el-message').isVisible()).toBe(false)
    })

    test('should called onClose', () => {
      let proxy
      const onClose = (componentInstance) => {
        proxy = componentInstance.proxy
      }
      const wrapper = mount(Message, {
        props: {
          onClose,
          showClose: true
        }
      })

      wrapper.get('.el-message__closeBtn').trigger('click')
      expect(proxy.close).toBeTruthy()
    })

    test('should close message when call close function in the componentInstance ', async () => {
      const wrapper = mount(Message)

      wrapper.vm.close()
      await flushPromises()
      expect(wrapper.get('.el-message').isVisible()).toBe(false)
    })

    test('should clear timeout', () => {
      jest.useFakeTimers()

      const wrapper = mount(Message, {
        props: {
          duration: 1000,
          showClose: true
        }
      })

      wrapper.get('.el-message__closeBtn').trigger('click')
      jest.runTimersToTime(1000)
      expect(wrapper.emitted('close')).toBeTruthy()
      expect(wrapper.emitted('close')[0][0].proxy.close).toBeTruthy()
    })
  })

  test('conter', () => {
    const wrapper = mount(Message, {
      props: {
        center: true
      }
    })

    expect(wrapper.get('.el-message').classes()).toContain('is-center')
  })

  test('customize the name of the class', () => {
    const wrapper = mount(Message, {
      props: {
        customClass: 'foo '
      }
    })

    expect(wrapper.get('.el-message').classes()).toContain('foo')
  })

  test('dangerouslyUseHTMLString', () => {
    const wrapper = mount(Message, {
      props: {
        dangerouslyUseHTMLString: true,
        message: "<div id='foo'>foo</div>"
      }
    })

    expect(wrapper.find('#foo').exists()).toBe(true)
  })

  test('verticalOffset', () => {
    const wrapper = mount(Message, {
      props: {
        offset: 50
      }
    })

    expectHaveStyle(wrapper.get('.el-message'), {
      top: '50px'
    })
  })

  test('update offset when called updateVerticalOffset', async () => {
    const wrapper = mount(Message, {
      props: {
        offset: 50
      }
    })

    expectHaveStyle(wrapper.get('.el-message'), {
      top: '50px'
    })
    wrapper.vm.offsetVal = 100
    await flushPromises()
    expectHaveStyle(wrapper.get('.el-message'), {
      top: '100px'
    })
  })

  test('should not closed when mouseenter ', async () => {
    jest.useFakeTimers()

    const wrapper = mount(Message, {
      props: {
        duration: 1000
      }
    })

    wrapper.get('.el-message').trigger('mouseenter')
    jest.runTimersToTime(1000)
    expect(wrapper.emitted('close')).toBeFalsy()
  })

  test('should  closed when mouseleave ', async () => {
    jest.useFakeTimers()

    const wrapper = mount(Message, {
      props: {
        duration: 1000
      }
    })

    wrapper.get('.el-message').trigger('mouseenter')
    wrapper.get('.el-message').trigger('mouseleave')
    jest.runTimersToTime(1000)
    expect(wrapper.emitted('close')).toBeTruthy()
  })
})
