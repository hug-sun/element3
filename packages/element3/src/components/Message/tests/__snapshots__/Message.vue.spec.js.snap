// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message.vue snapshot 1`] = `
<transition-stub>
  <div
    class="el-message"
  >
    <i
      class="el-message__icon el-icon-undefined"
    />
    
    <p
      class="el-message__content"
    />
    
    <!--v-if-->
  </div>
</transition-stub>
`;
